pkg: micropython: initial commit

kaspar030 opened the pull request:
This PR adds [Micro Python](https://github.com/micropython/micropython) as pkg (and a tests/ example).

It works on native, and I tried arduino due and iot-lab_M3. It should work on all 32bit patforms with enough (~12k, ~95k) RAM and flash.

The port doesn't do much as no libraries are included in the image.

- [x] the garbage collector doesn't consider register contents, so it will probably not work correctly
~~- [ ] there needs to be either a mapping for riot libs -> python libs or an adaption of micropython's "pyb." lib~~
~~- [ ] integrate with VFS~~ (postponed)

~~\- compile micropython for 16bit platforms depending on RIOT platform~~ (probably just needs too much memory)

UPDATE 07/2019:

- rebased to master
- rebased to current MicroPython master
- basic timer support using xtimer has been implemented
- machine.Pin() should work, if passed the right (integer) parameter as pin number
- added a basic test script

~~Needs #11841 for its ```git describe```-fix.~~ (#11841 is in)

OlegHahm commented:
awesome

kaspar030 commented:
I PR'ed this upstream: https://github.com/micropython/micropython/pull/1249

kaspar030 commented:
the "help()" problem somehow only shows with pyterm. Using picocom everything works fine.

OlegHahm commented:
pyterm handles `help` particularly because otherwise it would be handled by Python's cmd module. That may cause the problem.

kaspar030 commented:
@OlegHahm Lol! You know the feeling, staring at the debugger, debugging something that gets affected by debugging, but it should work, thinking, "WTF'ing hell is happening here!?" ;)

OlegHahm commented:
Sorry for that. I should looked at the todos...

phiros commented:
Needs rebase

authmillenon commented:
I'm still really excited for this! Let's try to get it in for 2016.07!

miri64 commented:
ping?

aabadie commented:
Any chance to see this in the next release ? (+1 on my side)

OlegHahm commented:
I heard somewhere that you could do cool stuff with (Micro)Python on RIOT...

kaspar030 commented:
alright, I'll dig this out.

- rebased
- updateded to latest upstream HEAD

aabadie commented:
@kaspar030, just gave a try to this PR and I issue at compile time. The option `shift-negative-value` is not supported by the version of GCC (5.4.0) I have (tested in native). I have the following error:
```error:cc1: error: -Werror=shift-negative-value: no option -Wshift-negative-value```

Removing `-Wno-error=shift-negative-value` fix the build issue on native.

I also tried with a samr21 board but then I have other build issues when applying the same trick:
```
CC RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S
arm-none-eabi-gcc -IRIOT/tests/pkg_micropython/bin/samr21-xpro/py_h/ -Wno-error=sign-compare -Wno-error=type-limits -Werror -mcpu=cortex-m0plus -mlittle-endian -mthumb -mfloat-abi=soft -mno-thumb-interwork -ffunction-sections -fdata-sections -fno-builtin -fshort-enums -ggdb -g3 -Os -std=gnu99 -fno-delete-null-pointer-checks -Wstrict-prototypes -Werror=strict-prototypes -Wold-style-definition -Werror=old-style-definition -fno-common -Wall -fshort-wchar -include 'RIOT/tests/pkg_micropython/bin/samr21-xpro/riotbuild/riotbuild.h' -DRIOT_VERSION=\"2016.07-devel-447-gb177-trek-saclay-micropython\" -DNDEBUG -I. -I.. -I../lib/mp-readline -I../stmhal -IRIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython -isystem /etc/alternatives/gcc-arm-none-eabi-include/nano -isystem /etc/alternatives/gcc-arm-none-eabi-include -IRIOT/core/include -IRIOT/drivers/include -IRIOT/sys/include -IRIOT/cpu/samd21/include -IRIOT/boards/samr21-xpro/include -IRIOT/cpu/sam21_common/include -IRIOT/cpu/cortexm_common/include -IRIOT/sys/libc/include -IRIOT/pkg/micropython/include -IRIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython -IRIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot  -Wno-vla -Wno-unused-parameter -Wno-pedantic -c -o RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/py/nlrthumb.o RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S: Assembler messages:
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:54: Error: cannot honor width suffix -- `str lr,[r0,#8]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:59: Error: cannot honor width suffix -- `str r8,[r0,#28]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:60: Error: cannot honor width suffix -- `str r9,[r0,#32]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:61: Error: cannot honor width suffix -- `str r10,[r0,#36]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:62: Error: cannot honor width suffix -- `str r11,[r0,#40]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:63: Error: cannot honor width suffix -- `str r13,[r0,#44]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:112: Error: cannot honor width suffix -- `ldr lr,[r2,#8]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:117: Error: cannot honor width suffix -- `ldr r8,[r2,#28]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:118: Error: cannot honor width suffix -- `ldr r9,[r2,#32]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:119: Error: cannot honor width suffix -- `ldr r10,[r2,#36]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:120: Error: cannot honor width suffix -- `ldr r11,[r2,#40]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:121: Error: cannot honor width suffix -- `ldr r13,[r2,#44]'
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:87: Error: invalid offset, value too big (0xFFFFFFFFFFFFFFF8)
RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/riot/../py/nlrthumb.S:104: Error: invalid offset, value too big (0xFFFFFFFFFFFFFFF0)
../py/mkrules.mk:25: recipe for target 'RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/py/nlrthumb.o' failed
make[2]: *** [RIOT/tests/pkg_micropython/bin/pkg/samr21-xpro/micropython/py/nlrthumb.o] Error 1
```

Do you know how to fix this ? Is it supposed to work on Samr21 (maybe there are limitations with this board) ?

kaspar030 commented:
> Removing -Wno-error=shift-negative-value fix the build issue on native.

Hmok, I changed the pkg Makefile to just compile all of MicroPython with ```-Wno-error```.

> Do you know how to fix this ? Is it supposed to work on Samr21 (maybe there are limitations with this board) ?

There seems to be some issue with MicroPython's assembler code for Cortex-M0(+). It is possible to compile MicroPython so it uses setjmp, which is probably less efficient, but at least it seems to compile & somehow work. I changed the PR accordingly.

aabadie commented:
Thanks @kaspar030. Tested on native, samr21-xpro, arduino-due => works !
The thing is that it provides a very limited version of micropython: for example, it doesn't provide access to the gpio. I guess we need some kind of HAL for this as micropython seems to be bound to the pyboard hardware.

kaspar030 commented:
@aabadie Yeah, that's why it is WIP.

aabadie commented on tests/pkg_micropython/Makefile:
```diff
@@ -0,0 +1,29 @@
+
+# If no BOARD is found in the environment, use this default:
+BOARD ?= native
+
```
I would suggest to add here (or somewhere it you think it's better) a list of unsupported boards, e.g 16bit and 8 bit boards.

aabadie commented:
@kaspar030, the HAL is a limitation and I guess it should be fixed/implemented upstream to make it RIOT compliant.
Do you think, this one can be merged as is or postponed until some more micropython libraries are available ?

miri64 commented on pkg/micropython/Makefile:
```diff
@@ -0,0 +1,12 @@
+PKG_URL=https://github.com/micropython/micropython
+PKG_VERSION=c3f519adfe1faf7808fa49fdb6006040317279ca
+
+PKG_SOURCE_OVERRIDE=/home/kaspar/src/micropython
```
I don't have this directory on my machine. Does this pose a problem?

kaspar030 replied:
> I don't have this directory on my machine. Does this pose a problem?

Ups, that's a leftover.

kaspar030 commented:
> @kaspar030, the HAL is a limitation and I guess it should be fixed/implemented upstream to make it RIOT compliant.

Well, I'm doing the upstream work. With a little luck, we can just map our API to micropython's C hal for basic stuff, but I'm afraid it'll not be that easy.

> Do you think, this one can be merged as is or postponed until some more micropython libraries are available ?

IMHO we should give it time until the next release. Until then, there'll be at least some functionality.

aabadie commented:
@kaspar030, did you make any progress on this one (and on the upstream fork) ?

miri64 commented:
Postponed due to Feature Freeze.

miri64 commented:
Any progress here?

drasko commented:
What's happening with this? Any updates?

miri64 commented:
Now that we have file system support (#5616) this should get some traction (?)

miri64 commented:
(oh, #5616 isn't merged yet but hopefully soon!)

miri64 commented:
For now we need to postpone however.

miri64 commented:
#5616 was merged. @kaspar030 any plans to work on this?

miri64 commented:
Ping?

aabadie commented:
Would be very nice to have this one merged soon. I discussed with someone from micropython project last week and he was very interested by this. What's missing exactly?

kaspar030 commented:
> What's missing exactly?

Actually, libraries. The interpreter works fine, but without any libs, it cannot do much. But jerryscript is in the same state, and it's been merged... I'll try to get back to this the next days in time for the release!

aabadie commented:
That would be awesome !

aabadie commented:
@kaspar030, do you think this PR could be merged in the current state (I mean without libraries) ?

smlng commented:
@kaspar030 do you have an ETA on this, feature freeze for 2018.01 is mid next week. I'll trigger Murdock in the mean time. If basic functionality like for `jerry/javascript` is working, we can merge if Murdock is happy. Otherwise we have to postpone this to 2018.04.

Agreed?

miri64 commented:
Ping @kaspar030? I rather go ahead and close this for now and give someone else the chance to do this if you do not have an ETA.

kaspar030 commented:
- updated to latest micropython

Still no ETA, but some progress.

tcschmidt commented:
@kaspar030 joining in with @miri64 ready for close?

tcschmidt commented:
Close due to long-term inactivity.

kaspar030 commented:
This is still on, don't close.

tcschmidt commented:
@kaspar030 well, I asked without an answer. Last real contributions in 2015 ... what is expected to happen here?

tcschmidt commented:
What are we going to do with this oldest PR ever? 
It would be much nicer to have microphython running on RIOT than just a 'historic' PR.
So I agree with @miri64 this should be taken up and worked on with fresh energy from the community. 
What about putting up an issue that calls for contributions to the porting of micropython?

emmanuelsearch commented:
@kaspar030 was trying to build this for iotlab-m3 but I get this error in machine_pin.c:

```
CC machine_pin.c
machine_pin.c: In function 'machine_pin_obj_init_helper':
machine_pin.c:63:17: error: 'GPIO_PUD' undeclared (first use in this function)
     uint pull = GPIO_PUD;
                 ^
machine_pin.c:63:17: note: each undeclared identifier is reported only once for each function it appears in
machine_pin.c: In function 'mp_pin_make_new':
machine_pin.c:91:17: warning: unused variable 'drv_name' [-Wunused-variable]
     const char *drv_name = mp_obj_str_get_str(items[0]);
                 ^
machine_pin.c: At top level:
cc1: warning: unrecognized command line option "-Wno-implicit-fallthrough"
/home/vagrant/RIOT/tests/pkg_micropython/bin/pkg/iotlab-m3/micropython/ports/riot/../../py/mkrules.mk:47: recipe for target '/home/vagrant/RIOT/tests/pkg_micropython/bin/pkg/iotlab-m3/micropython/machine_pin.o' failed
make[2]: *** [/home/vagrant/RIOT/tests/pkg_micropython/bin/pkg/iotlab-m3/micropython/machine_pin.o] Error 1
Makefile:11: recipe for target 'all' failed
make[1]: *** [all] Error 2
/home/vagrant/RIOT/tests/pkg_micropython/../../Makefile.include:369: recipe for target '/home/vagrant/RIOT/tests/pkg_micropython/bin/iotlab-m3/micropython.a' failed
make: *** [/home/vagrant/RIOT/tests/pkg_micropython/bin/iotlab-m3/micropython.a] Error 2
```

emmanuelsearch commented:
(and there micropython 1.9.4 is out now by the way ;)

x3ro commented:
I gave this a spin on the `stm32f429i-disc1` and the basic functionality (read: the REPL) works quite well. Thanks for the great work @kaspar030. Getting this merged would be great 🎉  Here's some feedback:

* It's a bit unexpected that one has to run `make prepare` manually. Can't we automate that?
* After a bit of tinkering I managed to get the `machine_pin` functionality to work (notes on this at the end), but maybe we could move `modmachine` into a separate PR to get this one merged quicker?
* If we're doing a sort-of fork from `micropython` for this (since it seems they won't add RTOS specific code to their main repo), shouldn't this live under `RIOT/` on GitHub?

---

### Regarding `machine_pin`

The following worked on the `stm32f429i-disc1` after changing `{.u_obj = MP_OBJ_NULL}` to `{.u_obj = mp_const_none}` and `uint pull = GPIO_OD_PU;` to `uint pull = GPIO_OUT;` in `machine_pin_obj_init_helper`:

~~~
Type "help()" for more information.
>>> import machine
>>> x = machine.Pin(("foo", 1073879054))
>>> x.init(machine.Pin.OUT)
>>> x.on()
~~~

Notable here is that the first argument of the tuple passed to `Pin` is unused. The second argument is the result of the `GPIO_PIN(PORT_G, 14)` macro of the `stm32_common` board (the red LED).

kaspar030 commented:
- rebased with #11841 merged.

Not waiting anymore!

jcarrano commented on pkg/micropython/Makefile:
```diff
@@ -0,0 +1,14 @@
+PKG_NAME=micropython
+PKG_URL=https://github.com/kaspar030/micropython
```
Could we have this pulled directly from upstream with our patches/contrib added on top?

Currently to know what has been changed I have to go to your fork and diff it against upstream.

kaspar030 replied:
> Could we have this pulled directly from upstream with our patches/contrib added on top?

Nope, that is a maintanance nightmare.

> Currently to know what has been changed I have to go to your fork and diff it against upstream.

Yes. ```make; cd bin/pkg/native/micropython; git diff master```.

Compare that to the extra work necessary to change only a single line in the micropython fork...

jcarrano commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,41 @@
+/**
```
Since the RIOT port seems to have several RIOT-specific additions, some more documentation would be welcome.

kaspar030 replied:
Please help out!

jcarrano replied:
No, man, this is _your_ feature.

jcarrano commented on tests/pkg_micropython/tests/01-run.py:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env python3
```
Does micropython have a more in-depth test suite? The reason I ask is that the changed introduced by the RIOT port seem to be quite extensive.

kaspar030 replied:
Possible. That can surely be added later. This test for now just confirms that the basic building is working and that the basic interpreter can execute simple python code.

jcarrano commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,41 @@
+ * ### thread_getpid()
+ *
+ *     $ import riot
+ *     $ print(riot.thread_getpid())
```
Would it be possible to map this to the `os` module instead of having RIOT-specific names?

kaspar030 replied:
Probably. There's no concept yet on how to map or integrate the (very pyboard specific) upstream API.

jcarrano commented on makefiles/py.inc.mk:
```diff
@@ -0,0 +1,19 @@
+	@mkdir -p $@
+
+$(PY_H): $(PY_HEADER_DIR)
+$(PY_H): $(PY_HEADER_DIR)/%.py.h: %.py $(PY)
```
Why does this depend on $(PY)

kaspar030 replied:
So it triggers a rebuild when the python files change.

jcarrano replied:
But why you want _any_ change in a python file trigger a rebuild in _all_ of them?

jcarrano commented on makefiles/py.inc.mk:
```diff
@@ -0,0 +1,19 @@
+# generate .py.h header files of .py files
```
This is tricky to get right while covering all (reasonable) use cases. Both Lua and Jerryscript leave it up to the user to define and  perform this step in the application Makefile.

kaspar030 replied:
> This is tricky to get right while covering all (reasonable) use cases.

Care to elaborate?

jcarrano commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,41 @@
+ *     $ t = xtimer.xtimer(lambda: print("timer triggered"))
+ *     $ t.set(100000)
+ *
+ * (this fails on ARM for lack of ISR stack)
```
I'd rather not include functionality that is known to be failing, at least not by default.

kaspar030 replied:
This fails for the same reason as why a printf() from a regular xtimer callback might fail. No reason not to include xtimer.

I can change the example to something like "change a global variable", but IMO that is much less conveying the point.

jcarrano commented on pkg/micropython/include/micropython.h:
```diff
@@ -0,0 +1,52 @@
+ * @param[in]   src pointer to Python code
+ * @param[in]   len length of src
+ */
+void mp_do_str(const char *src, int len);
```
Do strings get executed in the same environment/interpreter sessions, or is it a new session each time?

kaspar030 replied:
I've added a note (there's one global instance and the code will be executed in that instance).

kaspar030 commented on makefiles/py.inc.mk:
```diff
@@ -0,0 +1,19 @@
+# generate .py.h header files of .py files
+PY_HEADER_DIR ?= $(BINDIR)/$(MODULE)py_h
+PY = $(wildcard *.py)
```
Specifying the python modules manually is currently not possible, e.g., as "SRC_PY". The reason is that the application is built with "applications.mk", which overrides any application specific variables.

Consider the fail case (user script gets processed): a header will be generated in a non-clashing folder. IMO, that's acceptable.

Or do you have an example where a user had a python script in their application folder, and creating a header would break stuff?

This can be revisited once we have a more generic blob -> header solution.

kaspar030 commented on makefiles/py.inc.mk:
```diff
@@ -0,0 +1,19 @@
+
+# make C and C++ objects of this module depend on generated headers, so they
+# get re-build on changes to the python files.
+$(OBJC) $(OBJCXX): $(PY_H)
```
Chicken-and-egg problem: without specifying everything manually, the dependencies of a given ```.c``` are only available after building it. Without the header present, the build fails. Thus there's no (easy) way to get the dependency.

Making the objects of a given module that is using python files depend on the generated headers ensures they get re-built. It is possible that more objects than necessary get re-built. IMO that's acceptable for a first iteration, given that our makefiles basically require to "make clean" anyways.

kaspar030 commented on Makefile.base:
```diff
@@ -62,6 +62,8 @@ ASSMOBJ     := $(ASSMSRC:%.S=$(BINDIR)/$(MODULE)/%.o)
 OBJ := $(OBJC) $(OBJCXX) $(ASMOBJ) $(ASSMOBJ)
 DEP := $(OBJC:.o=.d) $(OBJCXX:.o=.d) $(ASSMOBJ:.o=.d)
 
+include $(RIOTMAKE)/py.inc.mk
```
Where would you put it? You mean "py.inc.mk" should be in "pkg/micropython"?

jcarrano replied:
I mean that now the rules for making the py.h  files are included regardless of whether micropython is being used or not.

This means that all python script are going to be processed always all the time. That is wat too invasive.

kaspar030 replied:
will add a guard

kaspar030 replied:
done

kaspar030 commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,41 @@
+ *
+ * MicroPython on RIOT has to be considered experimental. While the basic
+ * interpreter works fairly well on native an Cortex-M, it has not seen much
+ * testing. Especially GC and interrupt handling need some work.
```
It means that his is experimental, as in, there has not been much testing, and GC and interrupt handling will *definitely* have some issues. It means there are unknown unkowns.

jcarrano replied:
Do you have a plan to clear the unknowns?

kaspar030 replied:
I've removed the statement. GC should work fine (it is re-using upstream code) and interrupts are working fine (until proven otherwise).

kaspar030 commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,41 @@
+ * interpreter works fairly well on native an Cortex-M, it has not seen much
+ * testing. Especially GC and interrupt handling need some work.
+ *
+ * ## Available modules
```
no tests yet

kaspar030 replied:
added tests

kaspar030 commented:
@jcarrano thanks for reviewing!

This is considered a first iteration so we can collectively improve. Integrating Micropython properly has many steps, it is not feasible to do them all at once.

jcarrano commented:
> Integrating Micropython properly has many steps, it is not feasible to do them all at once.

Sure, that's why I insist on not including features instead of including things that don't work right (e.g. the Make integration). Otherwise we add up technical debt for the future that someone is going to have to fix at some point.

smlng commented:
question/suggestion: how about moving this port into its own repository, e.g. `RIOT-OS/micropython-riot`, with a fixed RIOT version a submodule (dependency)? With that we can improve the port step-by-step without messing with the RIOT-OS main repo?

aabadie commented:
Why not move the test application in examples ?

kaspar030 commented:
> Why not move the test application in examples ?

done

kaspar030 commented:
> question/suggestion: how about moving this port into its own repository, e.g. `RIOT-OS/micropython-riot`, with a fixed RIOT version a submodule (dependency)? With that we can improve the port step-by-step without messing with the RIOT-OS main repo?

Hm, the port has two sides:

1. integration in RIOT main tree, that is what this PR does
2. adding the port to upstream. The MicroPython side is currently located in a fork of upstream into my github name.

IMO, moving the micropython fork (2.) to RIOT-OS/micropython makes sense. Having a RIOT fork for 1. doesn't.

The port is actually in quite good shape. I managed to get the upstream test suite to execute (instructions included), as expected, most tests pass unless using modules that are not compiled in or using unavailable peripheral APIs. I'd say at this point this is as usable as jerryscript or lua.

smlng commented on Makefile.base:
```diff
@@ -62,6 +62,10 @@ ASSMOBJ     := $(ASSMSRC:%.S=$(BINDIR)/$(MODULE)/%.o)
 DEP := $(OBJC:.o=.d) $(OBJCXX:.o=.d) $(ASSMOBJ:.o=.d)
 
+ifneq (,$(filter micropython,$(USEPKG)))
+  include $(RIOTMAKE)/py.inc.mk
```
can this be moved to one of the Makefiles in `pkg/micropython/` ?

I'm not sure this is already answered, but generally it looks weird to have (single) package specific handling in here

kaspar030 replied:
I've factored this out into a generically usable and independently tested blob header: #11497.

OlegHahm commented:
So, what is the state of this one?

kaspar030 commented:
> So, what is the state of this one?

- basic functionality working, official test-suite can be run
- RIOT specific modules are documented
- open question: keep out of tree?
- this needs some kind of blob functionality. A simple (xxd) based variant (as used by jerryscript and lua) has been cut out into #11870.

kaspar030 commented:
>     open question: keep out of tree?

To quote myself, didn't get feedback on this:

> Hm, the port has two sides:
>
> - integration in RIOT main tree, that is what this PR does
>  - adding the port to upstream. The MicroPython side is currently located in a fork of upstream into my github name.
>
> IMO, moving the micropython fork (2.) to RIOT-OS/micropython makes sense. Having a RIOT fork for 1. doesn't.

aabadie commented:
> moving the micropython fork (2.) to RIOT-OS/micropython makes sense. Having a RIOT fork for 1. doesn't.

How about the 3rd option of merging the fork to micropython upstream ?

kaspar030 commented:
> How about the 3rd option of merging the fork to micropython upstream ?

That surely would be nice and should be our goal, but do you think we should postpone merging this integration PR until upstream is merged?

emmanuelsearch commented:
@aabadie since option 3 (upstream) is for later, what's your opinion on the options to proceed now?

aabadie commented:
> what's your opinion on the options to proceed now?

I would keep @kaspar030's fork because moving it to the RIOT-OS organization would give the message of a more "long term" setup, which is not expected IMHO. It's better to PR the changes required in fork to the upstream micropython repo. Once it's merged the target commit and url of the package can be updated easily.

kaspar030 commented:
- rebased, now that #11870 (blob support) is merged

bergzand commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,74 @@
+ * ## Implementation details
+ *
+ * The RIOT port of MicroPython currently resides in a fork at
+ * https://github.com/kaspar030/micropython. It is based on Micropython's
```
This should point to the [`add_riot_port`](https://github.com/kaspar030/micropython/tree/add_riot_port) branch

kaspar030 replied:
yup, fixed

bergzand commented:
I'm aware that it was requested before, but I find the move of the test application to the examples a bit odd. Does it make sense to have an example with an external python script (using the blob header integration) as an example and a separate test application with automated tests?

> I would keep @kaspar030's fork because moving it to the RIOT-OS organization would give the message of a more "long term" setup, which is not expected IMHO.

I think this depends a lot on how the "long term" development of the micropython port is going to look like. This of course also depends on the preference of the micropython developers. Either we swamp them with integration PRs for micropython with RIOT (think of all the `sys/` facilities offered) or we maintain a separate fork and upstream our integration features once so often. Maybe this is something we could discuss with their community at FOSDEM?

bergzand commented on examples/micropython/main.c:
```diff
@@ -0,0 +1,77 @@
+
+#include "blob/boot.py.h"
+
+static char mp_heap[MP_RIOT_HEAPSIZE];
```
Nitpick: For me the name here is not immediately clear, is there anything against simply naming it `MICROPYTHON_HEAPSIZE`? does the `mp_` prefix originate from micropython itself?

kaspar030 replied:
yes, micropython uses mp_ and MP_ everywhere. I don't care which we use...

bergzand replied:
Okay, lets keep it this way then for consistency.

bergzand commented on pkg/micropython/include/micropython.h:
```diff
@@ -0,0 +1,54 @@
+#endif
+
+#ifndef MP_RIOT_HEAPSIZE
+#define MP_RIOT_HEAPSIZE (16U*1024)
```
Could you add some documentation on this heap size? What value makes sense for an application? Maybe a pointer to micropython docs. Could be in the package doc.txt

bergzand commented on pkg/micropython/Makefile.dep:
```diff
@@ -0,0 +1,2 @@
+USEMODULE += xtimer
+USEMODULE += stdin
```
This file should contain the architecture blacklists nowadays if I understand correctly

kaspar030 replied:
yup, added avr8 and the ESPs for now.

bergzand replied:
ESPs or MSPs?

aabadie commented on examples/micropython/Makefile:
```diff
@@ -0,0 +1,42 @@
+BOARD_BLACKLIST += pic32-clicker pic32-wifire
+
+# boards with too little RAM or flash for MicroPython
+BOARD_INSUFFICIENT_MEMORY := blackpill bluepill i-nucleo-lrwan1 nucleo-f030r8 \
```
Should be moved to a separate Makefile.ci.

kaspar030 replied:
done

aabadie commented on examples/micropython/Makefile:
```diff
@@ -0,0 +1,42 @@
+
+# the MicroPython port depends on ISR_STACKSIZE, which is not available on
+# arm7, risc-v, pic32
+BOARD_BLACKLIST += avsextrem msba2
```
I prefer using `FEATURES_BLACKLIST += arch_xx` in the package dependency directly.

kaspar030 replied:
done

aabadie commented on examples/micropython/Makefile:
```diff
@@ -0,0 +1,42 @@
+BLOBS += boot.py
+
+# configure MicroPython's heap size
+CFLAGS += '-DMP_RIOT_HEAPSIZE=(8U*1024)'
```
Could be moved in the package `Makefile.include` file. This is not specific to the application, is it ?

kaspar030 replied:
it is here to show that it can be configured. if left out, there's a default as fallback.

aabadie replied:
IMHO, the best solution is to document it in the package documentation.

kaspar030 replied:
done

aabadie replied:
I'm not happy with the proposed solution: too low level, not well documented.

May I suggest to add the following in Makefile.include?
```mk
MP_RIOT_HEAPSIZE ?= 8182U

CFLAGS += -DMP_RIOT_HEAPSIZE=$(MP_RIOT_HEAPSIZE)
```
 (I don't think the quotes are required btw)

And in doc.txt:
```md
## Configuration options
 *
 * Use the following environment variables in the application Makefile
 * or from the command line to configure MicroPython:
 *
 * MP_RIOT_HEAPSIZE: heap size for Micropython. Defaults to 8KiB.
 *
 * Example on the command line:
 * ```
 * MP_RIOT_HEAPSIZE=16 make -C examples/micropython
 * ```
```

btw, the default is 8KiB, not 16.

kaspar030 replied:
done

aabadie commented on examples/micropython/main.c:
```diff
@@ -0,0 +1,77 @@
+         * reboots.
+         */
+        if (coldboot) {
+            printf("-- Executing boot.py\n");
```
```suggestion
            puts("-- Executing boot.py");
```

kaspar030 replied:
the compiler does that anyway, AFAIK even on -O0

aabadie replied:
I insist

aabadie commented on examples/micropython/main.c:
```diff
@@ -0,0 +1,77 @@
+        if (coldboot) {
+            printf("-- Executing boot.py\n");
+            mp_do_str((const char *)boot_py, boot_py_len);
+            printf("-- boot.py exited. Starting REPL..\n");
```
```suggestion
            puts("-- boot.py exited. Starting REPL..");
```

aabadie commented on examples/micropython/main.c:
```diff
@@ -0,0 +1,77 @@
+                }
+            }
+        }
+        printf("soft reboot\r\n");
```
```suggestion
        puts("soft reboot");
```
but why `\r` was needed ?

aabadie commented on examples/micropython/tests/01-run.py:
```diff
@@ -0,0 +1,84 @@
+    child.expect_exact('>>>')
+
+    duration = get_time() - before
+    assert(duration > 500)
```
```suggestion
    assert duration > 500
```

aabadie commented on examples/micropython/tests/01-run.py:
```diff
@@ -0,0 +1,84 @@
+        child.expect_exact('>>>')
+        duration = get_time() - before
+        print("test_sleep(%s, %s): slept %sms" % (t, slack, duration))
+        assert((duration > t) and (duration < (t + slack)))
```
```suggestion
        assert duration > t and duration < (t + slack)
```

aabadie commented on examples/micropython/main.c:
```diff
@@ -0,0 +1,77 @@
+{
+    int coldboot = 1;
+
+    /* let MicroPython know the top of this thread's stack */
```
Could the following code be moved to a shared function implemented in the package ?

kaspar030 replied:
yes. at the moment, the example contains a little more code than needed, some of it could be factored out into a shared function. But we don't know yet how MicroPython is gonna be used with RIOT. Will modules use it? Or will it be used for applications?

I want to work out those details as we go...

bergzand commented on examples/micropython/main.c:
```diff
@@ -0,0 +1,77 @@
+    mp_stack_set_top((char*)&stack_dummy);
+
+    /* Make MicroPython's stack limit somewhat smaller than actual stack limit */
+    mp_stack_set_limit(THREAD_STACKSIZE_MAIN - 128);
```
Maybe add a define for this 'somewhat smaller', maybe something along the lines of `MICROPYTHON_STACKSIZE_REDUCTION`.

kaspar030 replied:
ok, added to micropython.h

emmanuelsearch commented:
> Either we swamp them with integration PRs for micropython with RIOT (think of all the sys/ facilities offered) or we maintain a separate fork and upstream our integration features once so often. Maybe this is something we could discuss with their community at FOSDEM?

@bergzand @kaspar030 maybe, but I think we could have some of the discussion right here too, as @dpgeorge sometimes pops up in conversations around here too?

emmanuelsearch commented:
(or if this conversation promises to be too long, we can defer and adapt in subsequent PRs based on later conclusions)

kaspar030 commented:
> Either we swamp them with integration PRs for micropython with RIOT (think of all the `sys/` facilities offered)

IMO, there are three areas:

1. the interpreter itself
2. the upstream pyb.* API on top of the (Micro)Python standard library
3. and RIOT specific API mappings we come up with.

1.) should be in place already, as it is mostly platform independent code
2.) is *huge*. they have their own networking (based on lwip), their own "periph" (with different feature set), ... It'll be a lot of effort to get compatible
3.) will be easier, as we can freely add a subset of what RIOT offers at C level, without being tied to existing APIs or worrying about compatibility

Ideally, we'd have a full mapping / implementation of 2.), and we'd upstream that. I'm afraid that'll take a long time (and much work).
As an intermediate step, we should aim to make 3.) usable for more than "hello world" type applications, e.g., add simple gpio, timers (done), sock, coap wrappers.

We should discuss with upstream how interesting a port without or with only partial 2.) would be for them.

bergzand commented:
> Can you detail a bit more ?
> 
> IMHO, it's better to use a specific (release) commit of MicroPython in the RIOT package and replace ASAP the link to the fork with the link to the official repo. Which means patches required in MicroPython should be upstreamed.

I'm not against upstreaming our RIOT port for MicroPython, however I see some challenges in development as it is a bidirectional synchronization issue. With this port, only basic support for MicroPython on RIOT is offered while there are many more features that could be integrated into MicroPython such as peripherals, sock, gcoap, usb and saul. On the other hand there are synchronization issues between the repo as the micropython commit within our package needs to be synced with the RIOT release it is compatible with.

For this I see a RIOT specific fork of the MicroPython repo as one of the development strategies. With this we could first submit our own PRs there, review them ourselves and upstream our own changes in bulk to the MicroPython repo.

Our own MicroPython fork could be compatible with the last RIOT release and maybe up to master.
The work submitted upstream can then bring the Micropython port up to date with a current release of RIOT and the next MicroPython release would then be compatible with a specific RIOT release.

kaspar030 commented on examples/micropython/Makefile:
```diff
@@ -0,0 +1,42 @@
+RIOTBASE ?= $(CURDIR)/../..
+
+# the MicroPython port currently only supports 32bit architectures
+BOARD_BLACKLIST := arduino-duemilanove arduino-leonardo \
```
yup

bergzand commented:
Should we move this discussion to a separate issue and keep this PR focused on the PR?

aabadie commented on pkg/micropython/Makefile.dep:
```diff
@@ -0,0 +1,8 @@
+USEMODULE += stdin
+
+# MicroPython doesn't compile for <32bit platforms
+FEATURES_BLACKLIST += arch_8bit arch_16bit
```
```suggestion
FEATURES_REQUIRED += arch_32bit
```

kaspar030 replied:
that will require extra work when we support native 64bit, or fixup 16bit (which IIRC upstream should be able to do). Explicit is good!

bergzand commented on examples/micropython/Makefile:
```diff
@@ -0,0 +1,23 @@
+# select MicroPython package
+USEPKG += micropython
+
+# include blop.py as header
```
s/`blop`/`boot`/

aabadie commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,74 @@
+ *
+ * Steps:
+ *
+ * 1. flash tests/pkg_micropython
```
The Steps can be simplified in simple commands:
- for native
```
make -C examples/micropython all test
```
- on hardware:
```
make BOARD=samr21-xpro -C examples/micropython all test
```

kaspar030 replied:
No. The RIOT supplied test script tests the RIOT specific modules. The steps are for how to run the upstream test suite, which (currently) requires patching it in most cases, as RIOT's uart is not using flow control by default.

aabadie replied:
Ok, I see. But still, use `make -C examples/micropython flash` instead of `flash ...`. Better be explicit, right ?

Replace `tests/pkg_micropython` with `examples/micropython` in the other items.

bergzand commented on examples/micropython/main.c:
```diff
@@ -0,0 +1,77 @@
+ */
+
+/**
+ * @ingroup     tests
```
```suggestion
 * @ingroup     examples
```

aabadie commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,74 @@
+
+ * ### xtimer
+ *
+ *     $ import xtimer
```
I would replace the prompt character with the micropython one, e.g replace `$` with `>>>`. Same above for `thread_getpid()` example.

kaspar030 replied:
done

aabadie commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,74 @@
+ *
+ * ## How to use
+ *
+ * See tests/pkg_micropython for example code.
```
```suggestion
 * See examples/micropython for example code.
```

bergzand commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,81 @@
+ *     >>>
+ *     >>> t = xtimer.xtimer(inc_a)
+ *     >>> t.set(100000)
+ *     >>> print(a)
```
This example is broken for me:
```
inc_a()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in inc_a
NameError: local variable referenced before assignment
```

kaspar030 replied:
yup, it needs a "global a". fixed.

bergzand replied:
Ack, works now

emmanuelsearch commented:
> Should we move this discussion to a separate issue and keep this PR focused on the PR?

+1

dpgeorge commented:
Hi all, I'm happy to join the discussion about integration between upstream MicroPython and RIOT.  I'm not sure at this point what the right approach is, but, for the benefit of users, I would suggest that the Python-level API available on the RIOT port of MicroPython would at least contain a minimal subset of our "machine" API, with classes like Timer, ADC, SPI, UART, I2C being compatible.  This facilitates writing (Python level) drivers that run across everything.

aabadie commented on examples/micropython/Makefile:
```diff
@@ -0,0 +1,23 @@
+CFLAGS += '-DMP_RIOT_HEAPSIZE=(8U*1024)'
+
+# MicroPython needs a larger stack
+CFLAGS += '-DTHREAD_STACKSIZE_MAIN=THREAD_STACKSIZE_DEFAULT*4'
```
How do you determine this value ? Is it supposed to be configured by users ?
Same comment as above about moving this CFLAGS to the `Makefile.include` file of the package.

kaspar030 replied:
> How do you determine this value ? Is it supposed to be configured by users ?

It depends, as any stacksize value. This one, I doubled it until stack size related crashes stopped.

aabadie commented on pkg/micropython/Makefile.dep:
```diff
@@ -0,0 +1,8 @@
+# MicroPython doesn't compile for <32bit platforms
+FEATURES_BLACKLIST += arch_8bit arch_16bit
+
+# This port currently requires ISR_STACK and thread_isr_stack_start
```
```suggestion
# This port currently requires ISR_STACKSIZE and thread_isr_stack_start
```

aabadie commented on pkg/micropython/Makefile.dep:
```diff
@@ -0,0 +1,8 @@
+FEATURES_BLACKLIST += arch_8bit arch_16bit
+
+# This port currently requires ISR_STACK and thread_isr_stack_start
+FEATURES_BLACKLIST += arch_arm7 arch_esp32 arch_esp8266 arch_mips arch_riscv
```
arch_esp32 and arch_esp8266  provide ISR_STACKSIZE and thread_isr_stack_start:
```
$ git grep thread_isr_stack_start | grep esp
cpu/esp32/thread_arch.c:void *thread_isr_stack_start(void)
cpu/esp8266/thread_arch.c:void *thread_isr_stack_start(void)
$ git grep ISR_STACKSIZE | grep esp
cpu/esp32/include/xtensa_conf.h:#define ISR_STACKSIZE                 (2048)
cpu/esp8266/include/xtensa_conf.h:#define ISR_STACKSIZE                 (2048)
cpu/esp_common/vendor/xtensa/portasm.S:#define configISR_STACK_SIZE    ISR_STACKSIZE
```

kaspar030 replied:
the functions are guarded by DEVELHELP. I suggest fixing that in another PR, and not blocking this.

aabadie replied:
It's usable on hifive1b with this simple fix (not sure of what the value should be):
```diff
diff --git a/cpu/fe310/include/cpu_conf.h b/cpu/fe310/include/cpu_conf.h
index 698ea7873..a32d3c01d 100644
--- a/cpu/fe310/include/cpu_conf.h
+++ b/cpu/fe310/include/cpu_conf.h
@@ -34,6 +34,8 @@
 #endif
 /** @} */
 
+#define ISR_STACKSIZE                 (2048)
+
 /**
  * @brief   Declare the heap_stats function as available
  */
```

kaspar030 replied:
> It's usable on hifive1b with this simple fix (not sure of what the value should be):

nice! but the value is important here.

aabadie replied:
> the functions are guarded by DEVELHELP. I suggest fixing that in another PR, and not blocking this.

Yeah, I don't understand why it's done this way... Only one function (`int thread_isr_stack_usage(void)`) should be guarded.

aabadie replied:
> but the value is important here.

Indeed, I just reused the default value used for some other arch.

aabadie commented:
I found of couple of issues with basic features when testing the port:
- With pyterm, the Python prompt displayed contains spaces: `> > >`. When using socat, the prompt is correctly displayed. Maybe there's pyterm flags that should be added (I haven't checked which one and if there's one)
- Some nice MicroPython features cannot be used unless tweaking the terminal options. For example: the keyboard shortcut CTRL+D (soft reboot) exits pyterm/socat, enabling the paste mode with CTRL+E is unusable because you have to use CTRL+C (cancel) or CTRL+D (apply) once the code is pasted. A possible workaroung: add miniterm.py as optional RIOT terminal and force its use by default in the package configuration file.

kaspar030 commented:
> * With pyterm, the Python prompt displayed contains spaces: `> > >`. When using socat, the prompt is correctly displayed. Maybe there's pyterm flags that should be added (I haven't checked which one and if there's one)

aah. yup. I never use pyterm. It recognized ">" and handles it in a special way...

> * Some nice MicroPython features cannot be used unless tweaking the terminal options. For example: the keyboard shortcut CTRL+D (soft reboot) exits pyterm/socat, enabling the paste mode with CTRL+E is unusable because you have to use CTRL+C (cancel) or CTRL+D (apply) once the code is pasted. A possible workaroung: add miniterm.py as optional RIOT terminal and force its use by default in the package configuration file.

picocom also works fine. So what do we do. Maybe add a warning that suggests using a different terminal?

bergzand commented:
> picocom also works fine. So what do we do. Maybe add a warning that suggests using a different terminal?

I'm fine with a warning and a hint how to use a different terminal such as picocom.

aabadie commented:
I don't use picocom but miniterm.py is provided with pyserial which pyterm rely on. So that won't require to install another tool for people already used to pyterm. If we could have miniterm.py set as default `RIOT_TERMINAL` (just add `RIOT_TERMINAL ?= miniterm` in pkg/micropython/Makefile.include), that would do the trick nicely IMHO. Also remember #11003 ;)

kaspar030 commented:
> I'm fine with a warning and a hint how to use a different terminal such as picocom.

Hm, the warning get's printed first, and gets kinda drowned by all the compilation output...

kaspar030 commented:
> I don't use picocom but miniterm.py is provided with pyserial which pyterm rely on.

that's nice. So how about we keep the pyterm deault, move forward miniterm support, and change the example to use that once miniterm is merged?

kaspar030 commented:
> Hm, the warning get's printed first, and gets kinda drowned by all the compilation output...

I've pushed this:

if "term" is in make goals *and* RIOT_TERMINAL is pyterm, issue a warning, error out *unless* FORCE_PYTERM is 1.

emmanuelsearch commented:
+1 for pulling the pkg from a repo hosted by the RIOT organization as temporary solution for now, instead of @kaspar030 hosted repo.
(and let's indeed move longer term discussions in a different thread ;)

kaspar030 commented:
I've moved the micropython branch to RIOT-OS/micropython.

aabadie commented:
> I've pushed this:

Please consider #12743 instead of 11c4f4b

aabadie commented:
> +1 for pulling the pkg from a repo hosted by the RIOT organization as temporary solution for now, instead of @kaspar030 hosted repo.

As I said already in https://github.com/RIOT-OS/RIOT/pull/2968#issuecomment-529428519, I disagree with this. We should try to move as much code as possible to the MicroPython upstream repo.

aabadie commented:
I'll keep this PR blocked until 314da4a is removed.

aabadie commented:
> Some nice MicroPython features cannot be used unless tweaking the terminal options.

On native, these features are completely broken: Ctrl+D does nothing, Ctrl+C exits the application.

emmanuelsearch commented:
@aabadie I thought there was consensus on having the discussion on long term evolution of this in a separate thread? In the mean time, what is the show stopper with the current state?

bergzand commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,82 @@
+ * ## Implementation details
+ *
+ * The RIOT port of MicroPython currently resides in a fork at
+ * https://github.com/kaspar030/micropython (in branch add_riot_port). It is
```
Assuming 314da4a0b44a0566efc981761c6ee996bc332055 is going to stay, this should be updated.

kaspar030 replied:
yup

kaspar030 commented:
- rebased with miniterm merged, changed to use miniterm by default

kaspar030 commented:
> As I said already in [#2968 (comment)](https://github.com/RIOT-OS/RIOT/pull/2968#issuecomment-529428519), I disagree with this. We should try to move as much code as possible to the MicroPython upstream repo.

What's the problem? Merging the micropython as a pkg before the micropython side is upstreamed? Or the location of the fork?

aabadie commented:
Sorry, I should have quoted and commented the comment:

> I would keep @kaspar030's fork because moving it to the RIOT-OS organization would give the message of a more "long term" setup, which is not expected IMHO. It's better to PR the changes required in fork to the upstream micropython repo. Once it's merged the target commit and url of the package can be updated easily.

Here I meant that we should keep the url to @kaspar030's fork _until_ the required changes are upstreamed. Once this is done, the url of the package will simply have to be updated to the MicroPython repo url.
I really don't think a fork of MicroPython should be hosted under the RIOT-OS organization (see the first line of the quote).

aabadie commented on pkg/micropython/Makefile.dep:
```diff
@@ -0,0 +1,11 @@
+FEATURES_BLACKLIST += arch_8bit arch_16bit
+
+# This port currently requires ISR_STACKSIZE and thread_isr_stack_start
+FEATURES_BLACKLIST += arch_arm7 arch_esp32 arch_esp8266 arch_mips arch_riscv
```
```suggestion
FEATURES_BLACKLIST += arch_arm7 arch_esp32 arch_esp8266 arch_riscv
```

aabadie commented on pkg/micropython/Makefile.dep:
```diff
@@ -0,0 +1,11 @@
+FEATURES_BLACKLIST += arch_arm7 arch_esp32 arch_esp8266 arch_mips arch_riscv
+
+# The port currently doesn't compile for mips
+FEATURES_BLACKLIST += atch_mips32r2
```
```suggestion
FEATURES_BLACKLIST += arch_mips32r2
```

aabadie commented on pkg/micropython/Makefile.dep:
```diff
@@ -0,0 +1,11 @@
+# This port currently requires ISR_STACKSIZE and thread_isr_stack_start
+FEATURES_BLACKLIST += arch_arm7 arch_esp32 arch_esp8266 arch_mips arch_riscv
+
+# The port currently doesn't compile for mips
```
reason ?

kaspar030 replied:
multiple different errors, in addition to ISR_STACKSIZE missing

emmanuelsearch commented:
@aabadie @kaspar030 so how do we resolve here?

kaspar030 commented:
> @aabadie @kaspar030 so how do we resolve here?

move the pkg back to kaspar030/micropython.

emmanuelsearch commented:
makes sense to squash?

bergzand commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,82 @@
+ * ## Implementation details
+ *
+ * The RIOT port of MicroPython currently resides in a fork at
+ * https://github.com/RIOT-OS/micropython (in branch add_riot_port). It is
```
This should also be reverted again I guess.

aabadie commented on examples/micropython/Makefile:
```diff
@@ -0,0 +1,30 @@
+# characters (CTRL-D ...)
+RIOT_TERMINAL ?= miniterm
+
+# enable modmachine support for periphials if available
```
```suggestion
# enable modmachine support for peripherals if available
```

aabadie commented:
The content of this fixup commit (5760f5224c1ec66557b35124337dda0c2cc93053) doesn't correspond to the commit message. That might introduce weird things in the git history after squashing.

aabadie commented:
I still have a couple of unaddressed comment: move the CFLAGS assignments from the application Makefile to the Makefile.include of the package. The default RIOT_TERMINAL value could also be set there instead of the application Makefile.
Those configuration values could be document in the doc.txt of the package.

bergzand commented:
> The default RIOT_TERMINAL value could also be set there instead of the application Makefile.

I disagree with this. The `RIOT_TERMINAL` is only required for the case where the micropython REPL shell is used, as is the case with the example application here, but doesn't have to be the case in other situation such as automated tests. Changing the shell when the package is included seems a bit too early to me.

miri64 commented:
> > @aabadie @kaspar030 so how do we resolve here?
> 
> move the pkg back to kaspar030/micropython.

and from https://github.com/RIOT-OS/RIOT/pull/2968#discussion_r304838978:

> > Could we have this pulled directly from upstream with our patches/contrib added on top?
> 
> Nope, that is a maintanance nightmare.

May I ask how a orderly set of patches is a bigger maintanence nightmare than pointing to some "random guy's" fork? You still have to rebase the patches (aka commits) every time you update the package. Granted these are 3 commands (2 if you know what your doing and don't use the `rm`) more per update

```sh
git rebase -i upstream/master
```

vs.

```sh
git rebase -i upstream/master
rm *.patch
git format-patch upstream/master
cp *.patch ../RIOT/pkg/micropython/patches
```

but in the end it is way better understandable and transparent what modifications were done to the upstream right from the RIOT repository. Yes, there were reviewers being confused about patches in the past, because GitHub renders the change confusingly, but that is the only "nightmare" I can think of. This however, is more about personal habits than maintainability.

Another thing to consider is maintenance by other users: If I update the `pkg` with your approach I either have to fork it myself and then bend it over to my repo (giving the users no transparency about *my* changes) or I have to PR against your fork, putting making you the bottleneck for the new update still. With patches, any maintainer of `RIOT-OS` can review and merge them *and* everyone knows with one look (or more looks if the patch is bigger) what I did.

kaspar030 commented:
> May I ask how a orderly set of patches is a bigger maintanence nightmare than pointing to some "random guy's" fork

Well, because the workflow for doing changes with patches vs fork changes from:

1. check out fork
2. make change
3. pr/push
4. (reviewer) review in github, test by locally changing pkg commit
5. in case of change request, push fixups to pr
6. pr pkg commit change

to

1. checkout upstream
2. reset to correct base revision
3. apply patches
4. do change
5. create patch
6. pr patch to RIOT
7. (reviewer) repeat 1-3 for reviewing. comments only possible on patch.
8. in case of change request, push change commit to pkg branch, create new patch, push new patch to RIOT PR, reviewer has to manually apply additional patch, ...
9. when done, squash pkg branch, ...

Especially the whole github workflow is broken when using patches.
(

kaspar030 commented:
> I disagree with this. The `RIOT_TERMINAL` is only required for the case where the micropython REPL shell is used

this

kaspar030 commented:
- rebased, squashed

aabadie commented:
> May I ask how a orderly set of patches is a bigger maintanence nightmare than pointing to some "random guy's" fork

In the case of this PR, no patches would be required since the RIOT port is only adding adaption files to micropython upstream (in ports/riot). I have a branch ;) that adds the adaption code in a subfolder of the RIOT package and copy the these files on the fly in the upstream micropython ports/riot directory. That could work indeed and it's self contained.

This PR is based on a branch in @kaspar030's fork that is used in a PR to micropython upstream (https://github.com/micropython/micropython/pull/1249). Once the micropython PR is merged, we can just update the package URL/VERSION in RIOT to point to the upstream micropython url /commit.
In this case, if you want to extend the micropython port of RIOT, just PR directly to micropython, then update the RIOT package version.

aabadie commented:
https://github.com/RIOT-OS/RIOT/pull/2968/files#r347873773 is still unaddressed.

miri64 commented:
> Well, because the workflow for doing changes with patches vs fork changes from:

You could still have a fork (I used did this with `lwIP` e.g.) *and* use patches... Also idea is to keep the set of patches minimal of course. All major changes must either be provided to upstream or be provided in the adaption layer of the package. In my experience if you can't provide to upstream, can't do an adaption layer or have a huge, unmaintainable set of patches, you are doing something wrong.

Why does any other operating system find patches just fine to use with packages, while for RIOT according to you it is a maintenance burden?

haukepetersen commented:
> Why does any other operating system find patches just fine to use with packages, while for RIOT according to you it is a maintenance burden?

No idea, but IMHO patches are hell.

kaspar030 commented:
> You could still have a fork  _and_ use patches.

Yeah, to me, having the extra layer of pulling the source from github (or optionally a local folder) is much less work than dealing with patches, especially when collaborating. patches are *always* on top of what git/github already provides. Reviewing changes to patches is extremely cumbersome.

aabadie commented:
In case you missed https://github.com/RIOT-OS/RIOT/pull/2968#issuecomment-562075423, there's no need for patches in the micropython port. The contributed files are orthogonal with the current micropython code base: they are put in a separate subdirectory.

But I find it nice to contribute the RIOT adaption layer upstream: it makes it visible to micropython that a port for RIOT exists.

aabadie commented on pkg/micropython/doc.txt:
```diff
@@ -0,0 +1,93 @@
+ * interpreter works fairly well on native and Cortex-M, it has not seen much
+ * testing.
+ *
+ * ## Configuration options
```
This section should be removed now: it's duplicated with the new one added below.

kaspar030 replied:
yup. I removed the old one and moved the updated one up.

miri64 commented:
> > You could still have a fork  _and_ use patches.
> 
> Yeah, to me, having the extra layer of pulling the source from github (or optionally a local folder) is much less work than dealing with patches, especially when collaborating. patches are _always_ on top of what git/github already provides. Reviewing changes to patches is extremely cumbersome.

To me it is a matter of consistency and traceability. Let's say we have contributor `evilhaxxor` provide a package for a security relevant package. Which would you trust more? A package pulling upstream code for a dedicated version and a small set of well placed patches or a package that pulls in some commit from a fork of `evilhaxxor`? Yes you can go to the fork of `evilhaxxor` and look for the commit it is pointing to and the dig through the history to understand the changes, but that seems way more complicated then having the changes just there and reviewed in RIOT. 

I also don't get the PR argument. Your workflow would mean two PRs: One against the contributors fork and one against RIOT. Even worse, the PR against RIOT gives no immediate transparency about what the change is about. With patches just one PR is needed, so way lower maintenence burden for everyone involved + the patches and a well documented version jump give a better transparency about what is changed. 

But granted, I'm talking more about a stable package than a package currently under development like this one or `nimble` (/cc @haukepetersen), so maybe this discussion is a bit misplaced, but I have a problem with the generalized statement "patches are a nightmare/hell".

@aabadie, I read your comment and think it might be a good compromise until the RIOT code is merged upstream.

kaspar030 commented:
> But granted, I'm talking more about a stable package than a package currently under development like this one or `nimble` (/cc @haukepetersen), so maybe this discussion is a bit misplaced, but I have a problem with the generalized statement "patches are a nightmare/hell".

ok, then let me take back that general statement. In this case, we do expect quite some development work (until the changes are upstreamed), and the patches are quite substantial (>1300 lines spread over >20 commits), so I think in this case, we're better off with a (temporary) fork (and working with the patches would be a nightmare :wink:).

miri64 commented:
> ok, then let me take back that general statement. In this case, we do expect quite some development work (until the changes are upstreamed), and the patches are quite substantial (>1300 lines spread over >20 commits), so I think in this case, we're better off with a (temporary) fork (and working with the patches would be a nightmare wink).

Accepted. @aabadie's solution could also be a solution, but that would mean, once the changes *are* upstreamed they need to be removed again. Don't know what is worse :sweat_smile:.

aabadie commented:
Ideally the first step would be to revive micropython/micropython#1249, which just means remove WIP from the PR title and add a comment that this PR is ready. Maybe it could then be merged quite fast and the RIOT package could directly point to micropython upstream.

aabadie commented:
ACK when https://github.com/RIOT-OS/RIOT/pull/2968/files#r354255448 is addressed

kaspar030 commented:
> > The default RIOT_TERMINAL value could also be set there instead of the application Makefile.
> 
> I disagree with this. The `RIOT_TERMINAL` is only required for the case where the micropython REPL shell is used, as is the case with the example application here, but doesn't have to be the case in other situation such as automated tests.

@aabadie do you agree on this?

kaspar030 commented:
> Ideally the first step would be to revive [micropython/micropython#1249](https://github.com/micropython/micropython/pull/1249), which just means remove WIP from the PR title and add a comment that this PR is ready. Maybe it could then be merged quite fast and the RIOT package could directly point to micropython upstream.

That is parallel work. No need to wait with this PR. Also, I'd rather ask them to take another look at the upstream PR once all machine API's that @dpgeorge mentioned are actually implemented and when there's a way to compile from within MicroPython.

bergzand commented:
All green, GO!

kaspar030 commented:
> All green, GO!

Finally! :wink:

Thanks everyone involved!

OlegHahm commented:
Yeah! Awesome! Thanks for the great work.
